discovery.md

```diff
@@ -0,0 +1,378 @@
+producers. These take two forms:
+
+1. Middleware
+2. Aggregator (or *Events Broker*)
+
+In the Middleware case, discovery is a combined set of producers for where the
+discovery is passed through largely unmodified and subscriber URIs are generally
+absolute (back to the producer).
+
+In the aggregator use case, the aggregator is expected to modify the output of
+discovery and subscription URIs will be relative to the aggregator itself.
+
+## Notations and Terminology
+
+### Notational Conventions
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+### Terminology
+
+This specification defines the following terms:
+
+### Producer
+
+The "producer" is a specific instance, process or device that creates the data
+structure describing the CloudEvent.
+
+### Intermediary
+
+An "intermediary" receives a message containing an event for the purpose of
+forwarding it to the next receiver, which might be another intermediary or a
+consumer. A typical task for an intermediary is to route the event to receivers
+based on the information in the event context.
+
+### Middleware
+
+Specific to discovery and subscription capabilities. Events middleware systems
+provide discovery services that may be directly passing through discovery from
+producers, or may aggregate or otherwise alter the discovery. This includes
+events brokers that may also be relaying discovery AND subscription information
+further upstream to event producers. Multiple middleware systems may exist
+between the producer and the consumer.
+
+### Consumer
+
+The entity receiving the event - the destination for the events related to a
+Subscription when the events are “pushed” by the Event Producer. In some
+instances this might be the same entity as the Subscriber. It could be the
+eventual event consumer or a pub/sub messaging entity which delivers the event
+to the eventual event consumer.
+
+### Subscription
+
+The request for events from an Event Producer or middleware system.
+
+### Event Subscriber
+
+The entity managing the lifecycle of a Subscription on behalf of an Event
+Consumer. In some instances this might be the same entity as the Event Consumer.
+
+## API Specification
+
+The discovery API takes the declarative API design approach. This document
+structure focuses on discovery where each document has a two part key of
+`producer` and `type`. This takes a denormalized view where the composite key
+of documents is the `producer` and `type`. Producer is defined as an arbitrary
+human readable string and type is the CloudEvent type attribute.
+
+// TODO
+// 1) need to rationalize fanout, broker aggregator scenario
+// 2) Does discovery need to be 2 levels? - producers -> types -> sources
+
+### Producer entity Attributes
+
```

Comment:
It's interesting that producer becomes the first-class citizen for discovery, while we intentionally left it out of the event context for CloudEvents. I understand that you need something to enumerate on in addition to type, and that source is not suitable for this. Maybe I am really thinking too much about terminology, but could this be a _Provider_?

Reply:
Kind of related, one thing that's missing from the spec right now is how the data is returned? Is it just list where each entry is a flat list of attributes, or were people expecting some kind of object structure to it?  While not fancy, I'm partial to something that looks more like a table being returned because it's easy to process and also very easy to implement the search query - becomes a trivial SQL-like thing w/o concern for nested objects and how to express that (on input or output).

Reply:
My thought process here - going back to the discovery funnel [in the doc](https://docs.google.com/document/d/10S2MnYiy8wNBO8fXdtAviX0GvJXeoFqNOHG6Qh1-EAs/edit?ts=5e1767a6#heading=h.ihz1tqidrou8l)

As someone wanting to consume events, I see a list of "producers"  - systems that I can rationalize and understand.

Then I would select an event type from that producers

Then, if necessary, select a source from available sources for that producer + event type.